pinot-broker/src/main/java/com/linkedin/pinot/broker/queryquota/TableQueryQuotaManager.java

```diff
@@ -0,0 +1,265 @@
+    LOGGER.info("Rate limiter for table: {} has been initialized. Overall rate: {}. Per-broker rate: {}. Number of online broker instances: {}",
+        tableName, overallRate, perBrokerRate, onlineCount);
+  }
+
+  /**
+   * Acquire a token from rate limiter based on the table name.
+   * @param tableName original table name which could be raw.
+   * @return true if there is no query quota specified for the table or a token can be acquired, otherwise return false.
+   * */
+  public boolean acquire(String tableName) {
+    LOGGER.debug("Trying to acquire token for table: {}", tableName);
+
+    final String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+    String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(rawTableName);
+    String realtimeTableName = TableNameBuilder.REALTIME.tableNameWithType(rawTableName);
+
+    QueryQuotaConfig offlineTableQueryQuotaConfig = _rateLimiterMap.get(offlineTableName);
+    QueryQuotaConfig realtimeTableQueryQuotaConfig = _rateLimiterMap.get(realtimeTableName);
+
+    if (offlineTableQueryQuotaConfig == null && realtimeTableQueryQuotaConfig == null) {
+      LOGGER.debug("No qps quota is specified for table: {}", tableName);
+      return true;
+    }
+
+    // Try to acquire token for offline table if there's a qps quota.
+    if (offlineTableQueryQuotaConfig != null && !tryAcquireToken(tableName, offlineTableQueryQuotaConfig)) {
+      return false;
+    }
+
+    // Try to acquire token for offline table if there's a qps quota.
+    return realtimeTableQueryQuotaConfig == null || tryAcquireToken(tableName, realtimeTableQueryQuotaConfig);
+  }
+
+  /**
+   * Try to acquire token from rate limiter. Emit the utilization of the qps quota if broker metric isn't null.
+   * @param tableName origin table name, which could be raw.
+   * @param queryQuotaConfig query quota config for type-specific table.
+   * @return true if there's no qps quota for that table, or a token is acquired successfully.
+   * */
+  private boolean tryAcquireToken(String tableName, QueryQuotaConfig queryQuotaConfig) {
+    // Use hit counter to count the number of hits.
+    queryQuotaConfig.getHitCounter().hit();
+
+    RateLimiter rateLimiter = queryQuotaConfig.getRateLimiter();
+    double perBrokerRate = rateLimiter.getRate();
+    if (!rateLimiter.tryAcquire()) {
+      LOGGER.info("Quota is exceeded for table: {}. Per-broker rate: {}", tableName, perBrokerRate);
+      return false;
+    }
+
+    // Emit the qps capacity utilization rate.
+    if (_brokerMetrics != null) {
+      int numHits = queryQuotaConfig.getHitCounter().getHitCount();
+      int percentageOfCapacityUtilization = (int)(numHits * 100 / perBrokerRate);
+      LOGGER.debug("The percentage of rate limit capacity utilization is {}", percentageOfCapacityUtilization);
+      _brokerMetrics.addMeteredTableValue(tableName, BrokerMeter.QUERY_QUOTA_CAPACITY_UTILIZATION_RATE, percentageOfCapacityUtilization);
```

Comment:
You want to use `setValueOfTableGauge` here, since adding percentages doesn't really make sense.